### ISSUE TITLE
uproot.pandas.df(flatten=True) failing if varying dimensions of branches

### ISSUE THREAD

**Issue by sdporzio** [scikit-hep/uproot#227]
I've noticed uproot.pandas.df() smartly giving up on "flatten=True" for root files with branches that have dimensions >1 and use correctly "flatten=True" if dimensions = 0 or 1.

However for root files that have both (many branches with different dimensions), uproot first tried to load everything with flatten=True and then fails. This is easily solvable just forcing "flatten=False", but it might not be immediate for people encountering this problem.

**Comment by jpivarski** [comment]
I don't fully understand what you're saying, but it might be related to this recent change: https://github.com/scikit-hep/uproot/releases/tag/3.4.4

If all branches in the TTree are scalar-per-event or have the same number of elements as each other per event (e.g. there are different numbers of electrons in each event, but the same number of electron pT, electron eta, electron phi, ...), then `tree.pandas.df(flatten=True)` will work. If some branches have different numbers of elements than others per event, such as a mix of electron branches and muon branches, then there isn't a good DataFrame to represent that. Previously, we tried (by doing expensive "join" operations in the `tree.pandas.df` function), but recently we gave up on that for performance and sanity.

One solution, as you've said, is to force `flatten=False`, but this may not be what you want (Python lists in each DataFrame cell). Another solution, in the writeup I linked, is to select only compatible branches:

```python
tree.pandas.df(["MET_*", "Muon_*"], flatten=True)
```

Is this a request for better documentation? If so, where? I'm not sure where the users are and where they'll be searching when the naive `tree.pandas.df()` doesn't work. One thing I'm certain of is that they'll see the error message, and I tried to put the information there. Anyway, I'd want to encourage selective branch-reading, rather than `flatten=False`.

Or am I wrong and you were making a different point?

**Comment by chrisburr** [comment]
I think the suggestion is that the error message should be improved to say `Can't do that because....\n Do you want to pass "flatten=False"? See https://.... for details.`

**Comment by jpivarski** [comment]
Here's an attempt: https://github.com/scikit-hep/uproot/pull/226/commits/4551780479a7f23d0292e187604ae57a3de07a5a (I'm taking this onto an ongoing PR).

**Comment by sdporzio** [comment]
Yes, Chris is correct. Let me try to wrap up a bit better what I've been observing.

Trying to load up all the inconsistent branches (with different dimensions) gives the following error message:
```
df = rTree.pandas.df(['*'])
ValueError: cannot use flatten=True on branches with different jagged structure; explicitly select compatible branches (and pandas.merge if you want to combine different jagged structure)
```
which is pretty explanatory. Chris suggestion would definitely be an improvement over that.

However, what confused me is the fact that trying to manually select inconsistent branches (e.g. all `geo_*` branches with dimension 0 and all `cali_*` branches with dimension 3) gives:
```
df = a.pandas.df(['geo*','cali*'])
TypeError: NumPy boolean array indexing assignment requires a 0 or 1-dimensional input, input has 3 dimensions
```
which is a bit harder to point back to the flattening option.

**Comment by jpivarski** [comment]
Actually, that's why I asked for clarification: this looks like a bug. Can you show me the full stack trace?

**Comment by sdporzio** [comment]
Yes, of course
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-40-d29ca6464106> in <module>
----> 1 df = a.pandas.df(['geo*','cali*'])
      2 
      3 # df = a.pandas.df(['*'])

~/Software/Anaconda/lib/python3.7/site-packages/uproot/_connect/to_pandas.py in df(self, branches, namedecode, entrystart, entrystop, flatten, flatname, cache, basketcache, keycache, executor, blocking)
     48     def df(self, branches=None, namedecode="utf-8", entrystart=None, entrystop=None, flatten=True, flatname=None, cache=None, basketcache=None, keycache=None, executor=None, blocking=True):
     49         import pandas
---> 50         return self._tree.arrays(branches=branches, outputtype=pandas.DataFrame, namedecode=namedecode, entrystart=entrystart, entrystop=entrystop, flatten=flatten, flatname=flatname, cache=cache, basketcache=basketcache, keycache=keycache, executor=executor, blocking=blocking)
     51 
     52 def default_flatname(branchname, fieldname, index):

~/Software/Anaconda/lib/python3.7/site-packages/uproot/tree.py in arrays(self, branches, outputtype, namedecode, entrystart, entrystop, flatten, flatname, awkwardlib, cache, basketcache, keycache, executor, blocking)
    481         # if blocking, return the result of that function; otherwise, the function itself
    482         if blocking:
--> 483             return wait()
    484         else:
    485             return wait

~/Software/Anaconda/lib/python3.7/site-packages/uproot/tree.py in wait()
    465             import uproot._connect.to_pandas
    466             def wait():
--> 467                 return uproot._connect.to_pandas.futures2df(futures, outputtype, entrystart, entrystop, flatten, flatname, awkward)
    468 
    469         elif isinstance(outputtype, type) and issubclass(outputtype, dict):

~/Software/Anaconda/lib/python3.7/site-packages/uproot/_connect/to_pandas.py in futures2df(futures, outputtype, entrystart, entrystop, flatten, flatname, awkward)
    199                 fn = flatname(name, None, ())
    200                 array = awkward.numpy.array(array, dtype=object)
--> 201                 array = awkward.JaggedArray(starts, stops, awkward.numpy.empty(stops[-1], dtype=object))._broadcast(array).content
    202                 df[fn] = array
    203 

~/Software/Anaconda/lib/python3.7/site-packages/awkward/array/jagged.py in _broadcast(self, data)
    758             content[good] = data
    759         else:
--> 760             content[good] = data[self.parents[good]]
    761         out = self.copy(content=content)
    762         out._parents = self.parents

TypeError: NumPy boolean array indexing assignment requires a 0 or 1-dimensional input, input has 3 dimensions
```

**Comment by jpivarski** [comment]
I bet one of those arrays is empty. Is the file that causes this small enough to send to me? (Publicly or privately?) Thanks.

**Comment by sdporzio** [comment]
Yes, of course. I can send you the file. 
It should be available for a while at this link:
https://drive.google.com/open?id=1xc89zictutDHlbthLT_U7MoYKFOY5xYY

**Comment by jpivarski** [comment]
Thanks for the file. It was an unexpected behavior of Numpy: I was using a trick involving a Numpy object array, and it doesn't treat object arrays as opaque if the objects it's containing are Numpy arrays. So I worked around that case and all should work.

I put the fix into PR #226 for my own sanity—too many things to keep track of. It's therefore coupled to issue #225, but when that goes in, this will go in.

You might not be happy with the results. `pandas.df(flatten=True)` flattens only one level deep because the data stored in ROOT files is serialized differently when doubly nested. (I can't make the result generic without undoing that complexity by putting a lot more complexity into uproot.) So what you get is a DataFrame with a two-level MultiIndex ("entries" and "subentries") and the deeper elements are Python lists in the DataFrame cells. You can do a `df.apply` on that; I hope it's okay enough for what you're doing.

**Comment by sdporzio** [comment]
No, that sounds more than enough for the type of application I use.
Thanks for dealing with this so quickly!
